app/packages/looker/src/lookers/imavid/buffer-manager/index.ts

```diff
@@ -0,0 +1,216 @@
+import { BufferRange, Buffers } from "../../../state";
+
+/**
+ * Manages buffer ranges for ImaVid.
+ * Ranges are assumed to be inclusive, i.e. [start, end]
+ */
+export class BufferManager {
+  public buffers: Buffers;
+  private bufferMetadata: {
+    [rangeIndex: number]: string;
+  };
+
+  constructor(buffers: Buffers = []) {
+    this.buffers = buffers;
+    this.bufferMetadata = {};
+  }
+
+  /**
+   * Returns the total number of frames in the buffer.
+   */
+  get totalFramesInBuffer() {
+    return this.buffers.reduce(
+      (total, range) => total + range[1] - range[0] + 1,
+      0
+    );
+  }
+
+  /**
+   * Adds metadata to a buffer range at the specified index.
+   */
+  public addMetadataToBufferRange(rangeIndex: number, metadata: string) {
+    this.bufferMetadata[rangeIndex] = metadata;
+  }
+
+  /**
+   * Removes metadata from a buffer range at the specified index.
+   */
+  public removeMetadataFromBufferRange(rangeIndex: number) {
+    delete this.bufferMetadata[rangeIndex];
+  }
+
+  /**
+   * Returns the metadata for a buffer range at the specified index.
+   */
+  public getMetadataForBufferRange(rangeIndex: number) {
+    return this.bufferMetadata[rangeIndex];
+  }
+
+  /**
+   * Adds a buffer range to the buffer.
+   * If the new range overlaps with an existing range, the two ranges are merged.
+   * Time complexity: O(nlogn)
+   */
+  public addNewRange(
+    range: Readonly<BufferRange>,
+    ignoreRangesWithMetadata = true
+  ): void {
+    if (!range) {
+      return;
+    }
+
+    if (range[1] < range[0]) {
+      throw new Error("invalid range: range[1] must be >= range[0]");
+    }
+
+    // add the new range to the buffer
+    this.buffers.push(range);
+
+    // sort the buffers based on their start value
+    this.buffers.sort((a, b) => a[0] - b[0]);
+
+    // initialize a stack to store the merged ranges
+    const stack = [];
+
+    const rangesWithMetadata = [];
+    const rangesWithoutMetadata = [];
+
+    for (let i = 0; i < this.buffers.length; ++i) {
+      const range = this.buffers[i];
+
+      if (!range) {
```

Comment:
Maybe we can remove the safeguards before merging? `range` is not expected to be falsey, correct?

Reply:
It is not... but I got at least one instance where it was falsey and I couldn't reliably reproduce it, so until I figure that out, I'll leave it there.